### PR TITLE
Remove Synchronization from GameSequence.java  (was: Fix synchronization on non-final fields)

### PR DESCRIPTION
**Pull request by DanVanAtta**
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/708)

### CODE REVIEW

**Comment by RoiEXLab** [comment]
In ReadObject currentStepMutex is getting reset to an object... 
You'll either have to remove the final keyword, or remove the readObject code where a new object is created (I don't get what this is supposed to do)  or you could remove the initialization on declaration of you ensure that readObject is being called before the variable gets if readObject is really needed

**Comment by ron-murhammer** [comment]
@DanVanAtta Looks like the build is failing on this one?

**Comment by ron-murhammer** [comment]
@DanVanAtta Can you describe the reasoning behind this change? Is this causing a bug? Is it possible to have multiple instances of GameSequence?

File: src/games/strategy/engine/data/GameSequence.java [hunk 1]
```diff
@@ -14,7 +14,7 @@
   private int m_currentIndex;
   private int m_round = 1;
   private int m_roundOffset = 0;
-  private transient Object m_currentStepMutex = new Object();
+  private transient static final Object currentStepMutex = new Object();
```
Comment (DanVanAtta) [inline src/games/strategy/engine/data/GameSequence.java]: Static to protected against reflection not initializing this. There is only one GameSequence object, so sharing locks between multiple copies is okay.

**Comment by DanVanAtta** [comment]
> Can you describe the reasoning behind this change? 

Synchronization on non-final fields should be avoided.

> Is this causing a bug? 

None that I know of, but we do invite the opportunity.

> Is it possible to have multiple instances of GameSequence?

No, one per GameData.

**Comment by ron-murhammer** [comment]
@DanVanAtta But can't we have more than one GameData object? I don't think we want to make locks static if they don't need to be as its best to keep locking to a minimum for both performance and complexity sake.

**Comment by DanVanAtta** [comment]
Getting rid of this is a win:
```
-    if (m_currentStepMutex == null) {
 -      m_currentStepMutex = new Object();
 -    }
```
The above and moving to a final object reduces the complexity.

Yes we can have multiple GameData objects. Though, these methods really should not be competing with each other, and I'm not even sure where the multi-threading is coming from.. it is quite curious why they are synchronized at all. 

Regardless, all the operations they do are in-memory, so no performance hit.

**Comment by DanVanAtta** [comment]
Looking at GameSequence, there were a number of things wrong.
- The synchronization was not correct, one method was synchronized on the class, while the rest were synchronized on the mutex, they should all lock on the same object..
- One version of 'getStep' is synchronized, yet it calls another public method, another version of 'getStep', which is not synchronized. We have some code calling one, some calling the other, which would break the synchronization.
- Looking at the usages of each method, there is no multi-threaded shared state that would require synchronization.

**Comment by DanVanAtta** [comment]
- I peeled off the SwingComponents.java part of this update to: https://github.com/triplea-game/triplea/pull/752

So in this PR we can look at just the one other file with the synchronization removed

**Comment by ron-murhammer** [comment]
@DanVanAtta While I agree the synchronization is suspect here, I'm not sure whether we should remove it entirely. I'm guessing the reason it exists is because the engine can be updating the current round/step while things in the UI can potentially alter it as well. I would think this should be handled by locking the game data object rather than handling it at this low of a level. Maybe that is already done? I'm just a little hesitant to start playing around with the synchronization as it tends to be very difficult to test.

If you feel you spent the time evaluating that it isn't needed here then I'm ok moving forward. Just want to avoid potentially introducing bugs that are difficult to track down and fix.

**Comment by DanVanAtta** [comment]
I took a second look, this time reviewing the history of the code, and am still convinced the locking can be removed. IMO let's have a third look in a day or two, and double check the usages.


The original locking was added in here:
```
+/*
+ * GameSequence.java
+ *
+ * Created on October 14, 2001, 7:25 AM
commit ca34366b9a3865cf25362f630569c84db100c969
Author: sgbridges <>
Date:   Mon Feb 7 00:28:38 2005 +0000

    threading and ai fixes
```
My guesses for the locking was to prevent a remote host from changing the game round. Looks like we are dealing with with Swing which can spawn user events on multiple threads.  Perhaps the threading and AI fixes were related, related to how the AI was using this class, which might very well have been multi-threaded at the time.

The mismatching synchronization on the set method looks to have been done independently of the other locking, and was to support the 'save at a point in history' feature:

```
    Adding ability to specify which round the game starts on, through a round offset in gameplay section of xml. (veqryn)

git-svn-id: https://svn.code.sf.net/p/triplea/code/trunk/triplea@3882 a4c711aa-641f-0410-879d-9b464e52b682
+       public synchronized void setRoundAndStep(final int currentRound, final String stepDisplayName, final PlayerID player)
``` 

Trying out the save-in-history feature, the UI immediately goes to a prompt. So there is no way to generate multiple long-lived requests via the UI.

Meanwhile, the other synchronization is largely on 'getter' methods, and are between them. Another look is needed at this, but if all we are doing is locking in this class, then we still have a race condition. Specifically the blocked 'get' requests will just be blocked either behind or ahead of any data mutation. The caller will not know which, and hence there is a race condition.

So to recap, let's reconsider this once more to be sure.

**Comment by RoiEXLab** [comment]
@DanVanAtta Same here: Please resolve the merge conflicts

**Comment by ron-murhammer** [comment]
@DanVanAtta Thoughts on this one? I'm still a little hesitant to make synchronization changes especially close to the next release. But if you feel confident on it then I'm alright moving forward. I'd like to either close this or rebase and get it merged soon as its been open a while now.

**Comment by DanVanAtta** [comment]
Redoing these changes might make sense, too many conflicts to fight through. For the record, it does appear that there is a lot of unnecessary locking, which slows down the game and increased code complexity. Should be plenty of opportunities out there to remove unnecessary sync logic.
